src/storage/data_table.cpp

```diff
@@ -85,15 +85,97 @@ void DataTable::Scan(Transaction &transaction, DataChunk &result, TableScanState
+    T *max = (T *)state.column_scans[table_filter.column_index].current->stats.maximum.get();
+    switch (table_filter.comparison_type) {
+        case ExpressionType::COMPARE_EQUAL:
+		    return constant >= *min && constant <= *max;
+        case ExpressionType::COMPARE_GREATERTHANOREQUALTO:
+            return constant <=*max;
+        case ExpressionType::COMPARE_GREATERTHAN:
+            return constant <*max;
+        case ExpressionType::COMPARE_LESSTHANOREQUALTO:
+            return constant >=*min;
+        case ExpressionType::COMPARE_LESSTHAN:
+            return constant > *min;
+    default:
+            throw NotImplementedException("Operation not implemented");
+    }
+}
+
+bool DataTable::CheckZonemap(TableScanState &state , vector <TableFilter>&table_filters, idx_t &current_row){
+    bool readSegment = true;
+    for (auto & table_filter:table_filters) {
+        switch (state.column_scans[table_filter.column_index].current->type) {
+            case TypeId::INT8:{
+                int8_t constant = table_filter.constant.value_.tinyint;
+                readSegment &= checkZonemap<int8_t>(state, table_filter, constant);
+                break;
+            }
+            case TypeId::INT16:{
+                int16_t constant = table_filter.constant.value_.smallint;
+                readSegment &= checkZonemap<int16_t>(state, table_filter, constant);
+                break;
+            }
+            case TypeId::INT32:{
+                int32_t constant = table_filter.constant.value_.integer;
+                readSegment &= checkZonemap<int32_t>(state, table_filter, constant);
+                break;
+            }
+            case TypeId::INT64:{
+                int64_t constant = table_filter.constant.value_.bigint;
+                readSegment &= checkZonemap<int64_t>(state, table_filter, constant);
+                break;
+            }
+            case TypeId::FLOAT:{
+                float constant = table_filter.constant.value_.float_;
+                readSegment &= checkZonemap<float>(state, table_filter, constant);
+                break;
+            }
+            case TypeId::DOUBLE:{
+                double constant = table_filter.constant.value_.double_;
+                readSegment &= checkZonemap<double>(state, table_filter, constant);
+                break;
+            }
+            default:
+                throw NotImplementedException("Unimplemented type for uncompressed segment");
+        }
+        if (!readSegment){
+            //! We can skip this partition
+            idx_t vectorsToSkip = ceil((double)(state.column_scans[table_filter.column_index].current->count + state.column_scans[table_filter.column_index].current->start - current_row)/STANDARD_VECTOR_SIZE);
+            for (idx_t i = 0; i < vectorsToSkip; ++i){
+                //! nothing to scan for this vector, skip the entire vector
+                for (idx_t j = 0; j < state.column_ids.size(); j++) {
```

Comment:
The code here looks duplicated from DataTable::ScanBaseTable, could you move this to a separate function, e.g. ScanState::NextVector.

Reply:
I guess you meant TableScanState? I've created a nextvector method there